### ISSUE TITLE
Plugin Schema Issues 

### ISSUE THREAD

**Issue by cstones** [marcusgreen/moodle-qtype_gapfill#38]
Hello, we have this plugin in our system and the db/update.php doesn't
use separate code blocks for each new version. It seems the code is edited
in place rather than adding save point code blocks. This is making continuing
to use the plugin difficult.

As part of our upgrade process we also run all the Moodle Core Test scripts
to insure all data is maintained properly in the database.
Due to issues running the admin/cli/check_database_schema.php script
we are unable to update the plugin. 

The core script reports differences in the install.xml and upgrade.php
that have to be fixed.

The output from the script is as follows. 
```
$ php admin/cli/check_database_schema.php
-------------------------------------------------------------------------------
question_gapfill
 * column 'letterhints' should allow NULL (I)
-------------------------------------------------------------------------------
question_gapfill_settings
 * column 'question' has incorrect type 'C', expected 'I'
 * column 'itemid' has incorrect type 'C', expected 'X'
 * column 'gaptext' has incorrect type 'C', expected 'X'
 * column 'correctfeedback' should allow NULL (X)
 * column 'incorrectfeedback' should allow NULL (X)
-------------------------------------------------------------------------------
```
Could these issues be address? Do you use the xmldb editor for creating upgrade code?

We very much appreciate all the effort and on going efforts
with this plugin!

**Comment by marcusgreen** [comment]
I was only looking at that upgrade code today for the purpose of upgrading my wordselect question type.  The way it works doesn't adhere to the standard but because it checks for the existence of tables before making changes it continues to work. I am not due to make any table structure changes in the near future and I am not aware of any advantage in retrofitting the standard version based way of doing it. If you have an issue with that I can certainly look into doing that retrofit.
I considered myself the king of the Moodle database structure e.g.
http://www.examulator.com/er/
Until I received your report :). Until very recently I did not use the xmldb editor but it is now my way of doing things, and your report on the types illustrates why it is important.

My next intended upgrade to the Gapfill question type is planned to be a slight modification to the way the Javascript works (AMD modules), which will also allow it to work in activities other than the core quiz. Please email me directly at marcusgreen at mailbox.org so I can co-ordinate the next upgrade with you and ensure all your items are addressed.

**Comment by marcusgreen** [comment]
It was late at night when I posted my response and I have just had a closer look at the XML code here and I noticed that the only notnull columns are 
id, and question which is correct as there would be no meaning in having a record without one matching in the question table.  

Also I don't know what it means by types C and X. ? When I ran the check_database_scheme.php file it didn't give any output relating to the gapfill question type, even though it was installed. I have done some web searches on those output strings but have not found any explanations of those errors. Could you confirm the version of Moodle being used?.

**Comment by cstones** [comment]
We can only upgrade based on the upgrade.php code. In your testing are you able to drop the existing tables created from install.xml and then force run the upgrade code instead to reproduce the issues we see?

**Comment by marcusgreen** [comment]
I have been making changes tonight to upgrade.php so the code is generated by the dbxml editor. You should be able to see the updates here
https://github.com/marcusgreen/moodle-qtype_gapfill/commits/master/db/upgrade.php
I have not tested this yet I will post an update here when I have done that.

**Comment by marcusgreen** [comment]
A fix to this issue will be in the next release. I try to credit anyone who contributes to the development in this file
https://github.com/marcusgreen/moodle-qtype_gapfill/blob/master/changelog.md
Can I mention that you spotted and reported this in the release notes, what name/organisation should I use?

**Comment by marcusgreen** [comment]
Have just released v1.975 to the Moodle.org plugins database which includes this fix so am closing this issue. Many thanks again for reporting it.
